unit_testing/scripts.py

```diff
@@ -0,0 +1,63 @@
+# helper module to expose cli scripts outside top-level package
+
+import os
+import sys
+
+from spinalcordtoolbox import __sct_dir__
+
+# add scripts location to path
+sys.path.append(os.path.join(__sct_dir__, 'scripts'))
+
+# import all scripts
+import sct_analyze_lesion
```

Comment:
I'm a little confused about why we're doing the imports here in this file. If we're adding `scripts/` to the path, can't we just do `from scripts import <sct_function>` directly in each test?

Reply:
To avoid adding it to the path in each test file. Tests can be run individually therefore they cannot assume that `scripts` is in the path since it is not part of the package.

We can't do `from scripts import <sct_function>`, because we are adding the `scripts` directory itself to the path. Only the contents (i.e the scripts themselves) is accessible.

What I like about the proposed approach here is that it is explicit and straightforward. `scripts.py` exposes the scripts and the tests that need to import a script may just use it via relative import of `.scripts`.

Reply:
> To avoid adding it to the path in each test file. Tests can be run individually therefore they cannot assume that scripts is in the path since it is not part of the package.

Would `conftest.py` be an appropriate place to put this to cover all of the tests? Actually... it's already there. :sweat_smile: 

https://github.com/neuropoly/spinalcordtoolbox/blob/c8eaa4a202deaaba72f080c7506e5bb0b144efa6/conftest.py#L23-L25

> We can't do from scripts import <sct_function>, because we are adding the scripts directory itself to the path. Only the contents (i.e the scripts themselves) is accessible.

Huh... this is a bit confusing to me. I tried `from scripts import <sct_function>` for a test and it worked? Perhaps I am misunderstanding something about Python imports.

Scripts currently has an `__init__.py` thanks to #2837 so I believe it's treated as its own package... perhaps that's why it's working for me?

> What I like about the proposed approach here is that it is explicit and straightforward. scripts.py exposes the scripts and the tests that need to import a script may just use it via relative import of .scripts.

My only concern is that it requires manual updating when a script is added/removed/etc. But apart from that, I agree.

Reply:
> Would `conftest.py` be an appropriate place to put this to cover all of the tests? Actually... funny enough, it's already there. So, currently the scripts folder is added to the path twice.

It's there, but it shouldn't be. It's only there because `sct_download` is used in a pre-test collection hook and we haven't yet decided what the final form of `sct_download` should be.

> Huh... this is a bit confusing to me. I tried from scripts import <sct_function> for a test and it worked? Perhaps I am misunderstanding something about Python imports.

Probably because `$SCT_DIR` was added to `$PATH` somewhere else coincidentally. `scripts` doesn't have a parent package because the top-level `spinalcordtoolbox` folder doesn't have an `__init__.py`, otherwise we would have to do `from spinalcordtoolbox.scripts import sct_something` and `from spinalcordtoolbox.spinalcordtoolbox import some_api_module`.. If I launch a python shell in my venv and print `sys.path` I can see the `$SCT_DIR` in there whereas my regular path doesn't have it. In any case it is not the desired behavior.

> Scripts currently has an __init__.py thanks to #2837 so I believe it's treated as its own package... perhaps that's why it's working for me?

That's good, it allows a script to import another relative script via `from . import sct_something`. Same logic as what I am introducing in the `unit_testing` folder in this PR.

> My only concern is that it requires manual updating when a script is added/removed/etc

Yes and same goes for any API module once we merge https://github.com/neuropoly/spinalcordtoolbox/pull/2955/commits/f65df321b078b939e5de66755ca0561fe773ed82 because it allows more sane imports like `import spinalcordtoolbox as sct` with access to all sub-packages without explicit importing. For example you expect to do:
```
import os
os.path.join(...)
```
instead of:
```
import os
import os.path
import os.path.join
os.path.join(...)
```
Right now on `master` we can only do the second form, this PR combined with #2955 allows the first form which is standard practice.

Reply:
Thanks for the clarification. :slightly_smiling_face:

Reply:
The `os` example I gave is probably a bad one since `os` is part of the standard lib with an implementation scattered across multiple files. The `import numpy as np` and usage of `np` in general is a better one (https://github.com/numpy/numpy/blob/master/numpy/__init__.py)